Issue #14910: Add new section 4.8.5.2 in google style guide implementation

Zopsss opened the pull request:
Issue #14910 

~~Draft PR~~ because I had some doubts regarding failing xdoc tests.

I was adding a new section `4.8.5.2` to the google style guide implementation, to add support for this section, I required to combine 2 different modules ( AnnotationLocation & InvalidJavadocPosition ) in single test case. 

The test case is working fine but the problem is with Xdoc tests. I added 2 new configs and 1 test case in the coverage page, something like this:

![Screenshot 2024-06-02 012656](https://github.com/checkstyle/checkstyle/assets/86093112/c9037dce-b2b1-4f2c-a37e-1ba808a485a6)


Here, 2 `config` are for AnnotationLocation module and InvalidJavadocPosition module. The single `test` is the test file which uses these 2 configs internally to ensure we're following the rule.

But the `XdocPagesTest` is failing, it requires all the `config` to be paired with their individual `test` which is not the case here. In this case, it requires to have one `test` for AnnotationLocation and other `test` for InvalidJavadocPosition. 

And also it requires to have the test file linked in the `test` to be same name as its module name. In this case, for AnnotationLocation's `test`, it requires to be linked to a `AnnotationLocationTest` file and for InvalidJavadocPosition's `test`, it requires to be linked to a `InvalidJavadocPositionTest` file. In short, we cannot name the test file differently from its module name.


#### What we need for the new Xdoc model:

- Remove the pairing of `config` and `test` and allow us to have single `test` for multiple `config`. 
- The name of the test file used in the `test` does not have to be same as its module name.

romani commented:
```
[ERROR] Failures: 
[ERROR]   XdocsPagesTest.testAllStyleRules:1651->validateStyleModules:1810 
google_style.xml rule '4.8.5.2 Class Annotations' module 'AnnotationLocation' 
has mismatched config/test links
expected: test
but was : config

```

romani commented:
yes, you just need to update xdoc validation XdocsPagesTest.testAllStyleRules . 
section `4.8.5.2 Class Annotations` should be `ClassAnnotationsTest` or as before each Check has special test (to keep compatibillity with old style).
So is test is notmatching a Check, we check that xxxxTest  is matching title of section, without digits and special symbols (`-`, `:`) and it should be camel case.
please update this test in this PR, do not hesitate to change code of this test.

Zopsss commented:
How can I solve this error:

https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/49945078/job/rruisaqwb5xrrwyq#L311

It is for following line:

https://github.com/checkstyle/checkstyle/blob/33f16c552746ef1a6af9cbcc6c94826df776acdc/src/it/java/com/google/checkstyle/test/chapter4formatting/rule4852classannotations/ClassAnnotationsTest.java#L51

I tried using `DefaultConfiguration` and `Configuration` both as the variable type but it is still giving the same error

rdiachenko commented:
> How can I solve this error:
> 
> https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/49945078/job/rruisaqwb5xrrwyq#L311
> 
> It is for following line:
> 
> https://github.com/checkstyle/checkstyle/blob/33f16c552746ef1a6af9cbcc6c94826df776acdc/src/it/java/com/google/checkstyle/test/chapter4formatting/rule4852classannotations/ClassAnnotationsTest.java#L51
> 
> I tried using `DefaultConfiguration` and `Configuration` both as the variable type but it is still giving the same error

It is related to a method used to create config (not a type): `createTreeWalkerConfig(...)`. More details in [this javadoc](https://github.com/sevntu-checkstyle/sevntu.checkstyle/blob/master/sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/design/CheckstyleTestMakeupCheck.java#L52).

I see the same creation method used in `RightCurlyTest`, and I also see it is suppressed in `sevntu-suppressions.xml` config:
```
  <!-- Suppressed alternative method of pulling partial configuration from file -->
  <suppress checks="CheckstyleTestMakeup" files="[\\/]RightCurlyTest\.java"/>
```

You should probably extend that suppression for your test.

Zopsss commented:
> You should probably extend that suppression for your test.

thanks, now it is working fine. 

CircleCI CIs are failing:

https://app.circleci.com/pipelines/github/checkstyle/checkstyle/25711/workflows/82fbba72-a93e-4413-96fd-257f97beca74/jobs/598142

https://app.circleci.com/pipelines/github/checkstyle/checkstyle/25711/workflows/0b2da461-e9a2-4ba0-9cf5-0b7d146db772/jobs/598145

But I guess we do not need to worry about it. Maybe these are the reasons:

> CircleCI Security Alert:  We have become aware of a website impersonating CircleCI, presumably for malicious purposes. Stay alert and stay safe. [Read More](https://discuss.circleci.com/t/circleci-security-alert-warning-fraudulent-website-impersonating-circleci/50899)

> CircleCI ran into an issue setting up the infrastructure for this job. Try re-running the build and if this issue persists, open a [Support ticket](https://support.circleci.com/hc/en-us).

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -226,6 +226,61 @@ public class XdocsPagesTest {
 
+    // Contains all the sections which are not migrated to chapter wise testing
+    // in google style documentation.
+    // This list will be removed once all the sections are migrated.
```
please put here `until #xxxxxx`
to make sure in scope of what ticket we will do final cleanup.

Zopsss replied:
Currently, we does not have any issue which talks about migrating to chapter wise testing, so which issue should I link here?

romani replied:
lets create such issue and reference it here :)

Zopsss replied:
I have opened the issue: https://github.com/checkstyle/checkstyle/issues/14937. I'll link this as you requested.

Zopsss replied:
Linked the issue: https://github.com/checkstyle/checkstyle/blob/7e868f5cd0424fa926ed50a7426158666c3941a1/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L229-L233

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1773,6 +1828,93 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
         final Iterator<Node> itrChecks = checks.iterator();
         final Iterator<Node> itrConfigs = configs.iterator();
+        final boolean isRegularDocumentation = "google".equals(styleName)
+                && PER_MODULE_TESTS_RULES_LIST.contains(ruleName) || "sun".equals(styleName);
```
```
final boolean isPerModuleTesting = "sun".equals(styleName) 
   ||  "google".equals(styleName) && PER_MODULE_TESTS_RULES_LIST.contains(ruleName) 
```

Zopsss replied:
done

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1773,6 +1828,93 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
+                && PER_MODULE_TESTS_RULES_LIST.contains(ruleName) || "sun".equals(styleName);
+
+        if (isRegularDocumentation) {
+            while (itrChecks.hasNext()) {
```
please move to new method - validateModuleWiseTesting

Zopsss replied:
Done.

https://github.com/checkstyle/checkstyle/blob/7e868f5cd0424fa926ed50a7426158666c3941a1/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1835-L1837

https://github.com/checkstyle/checkstyle/blob/7e868f5cd0424fa926ed50a7426158666c3941a1/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1847-L1848

romani reviewed (changes requested):
items:

Zopsss commented:
@romani  I've made changes according to your suggestions. The XdocPagesTest now has support to check Chapter Wise Testing approach. I've tested it locally with different inputs and it was working fine.

 I guess the PR is now ready to be merged.

romani commented on config/sevntu-suppressions.xml:
```diff
@@ -39,6 +39,7 @@
 
   <!-- Suppressed alternative method of pulling partial configuration from file -->
   <suppress checks="CheckstyleTestMakeup" files="[\\/]RightCurlyTest\.java"/>
+  <suppress checks="CheckstyleTestMakeup" files="[\\/]ClassAnnotationsTest\.java"/>
```
It was checkstyle violation.
`ERROR] [checkstyle] [ERROR] C:\projects\checkstyle\src\it\java\com\google\checkstyle\test\chapter4formatting\rule4852classannotations\ClassAnnotationsTest.java:51:46: DefaultConfiguration was not assigned by an allowed create configuration method. [CheckstyleTestMakeup]`

When I removed such suppression, `mvn verify` passed, so it means this suppression is not required.

romani replied:
Let's try to remove it to let me see this anomaly

Zopsss replied:
done

Zopsss replied:
Failing CI: https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/49952243/job/96b020g8euk1q1ld#L311

romani replied:
It is weird why only windows is affected, Linux execution is green.
We will deal with this in separate issue.

romani replied:
Please restore suppression.

Zopsss replied:
done

rnveach replied:
Violation looks correct. `createTreeWalkerConfig` is not an allowed method. You can see it is extremely rare for us to use it in our code base.

I mentioned before this method of picking and pulling configs just doesn't seem like the best approach. How will we validate the checks we manually pick are the same ones we have identified in the sections and that we didn't miss one or add one? The whole reason we had a 1 to 1 check/test is this was a way to validate we were testing against what we are saying.

Is there a reason we just don't apply the whole google config against the input file?

romani replied:
https://github.com/checkstyle/checkstyle/issues/14939

rnveach replied:
My comment is not on the inconsistent CI, but on the code itself.

romani replied:
>Is there a reason we just don't apply the whole google config against the input file?

It might be too much violations, or too hard to make Input.
I remember we tried at initial year to make it and there was some mess. 
I am not sure on this.
It is easy to make 0 violations Inputs, and with violations might have bunch unrelated violations.

Zopsss replied:
> How will we validate the checks we manually pick are the same ones we have identified in the sections and that we didn't miss one or add one? The whole reason we had a 1 to 1 check/test is this was a way to validate we were testing against what we are saying in the documentation.


You have a valid point. We cannot test if all the mentioned modules in a section are used in the test or if any module is missing in the test. This approach focuses on combining different modules into one test so we can test the whole section in one go. For this, I guess we need to manually ensure that all the modules are valid and no modules is left behind.

rnveach replied:
Manual is ok for a time, but I see this as being a concern. All reviewers must ensure every tests matches google doc each time an update is done.

I don't see how it is bad we have unrelated violations if we apply the whole google config against input files. The whole point of this is we are showing we follow google's guide, so any and all violations which are displayed should be part of what we are showcasing. If a violation is "wrong", then this would be a bug in our config and would need to be fixed.

Zopsss replied:
>  My motto is if you don't enforce something by CI, someone will find a way to do something the wrong way. 

Why CI? We can write test in XdocPagesTest so why we should enforce something by CI?

> I don't see how it is bad we have unrelated violations if we apply the whole google config against input files. The whole point of this is we are showing we follow google's guide, so any and all violations which are displayed should be part of what we are showcasing. If a violation is "wrong", then this would be a bug in our config and would need to be fixed.

You're correct but checking the input against whole config will require us to take small things into account like adding javadocs, maintaining indentation ( google uses 2 spaces for indentation ), etc. This can become annoying to do as we need to ensure all these small details in every tests we write. Romani also said in the discord that, earlier we tried this approach but it was a mess or something like that. IMO we should properly discuss this before moving forward. I'm fine with abandoning the idea of validating the modules being used in the test. We can implement it after we come to a solid decision.

rnveach replied:
> We can write test in XdocPagesTest 

Tests are run by CI, so if it is in a test, it is enforced. I wasn't implying writing in the CI's YML directly, sorry for that confusion.

> checking the input against whole config will require us to take small things into account like adding javadocs, maintaining indentation ( google uses 2 spaces for indentation ), etc. This can become annoying to do as we need to ensure all these small details in every tests we write. 

Yes, this is why I was suggesting this not be part of this GSoC. It does also seem odd to show these files off as following GSoC, when as you bring up, they do not follow them. :)

> I'm fine with abandoning the idea of validating the modules being used in the test. We can implement it after we come to a solid decision.

I agree.

Zopsss commented:
GitHub, generate site

romani commented:
@Zopsss , please create issue to discuss and do xdoc pages validation for test presense.
in this PR do hardcodes list of sections where we simply skip validation of tests in table.
We will discuss all details on how to do this in separate Issue of this summer project.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1818,34 +1891,111 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
+            final boolean moduleIsCheck = href.startsWith("checks/");
+
+            if (!moduleIsConfig && !moduleIsCheck) {
+                skip = true;
```
In master, this type of condition was hit on a link to an issue ( `https://github.com/checkstyle/checkstyle/issues/265` ).

rnveach replied:
I am seeing separation a bit better now, skip will skip all test config links if it is set. This could lead to false results where we completely ignore column despite what is in check column.

If column should be empty, it should be validated as empty. If it shouldn't be empty, it should be validated as not empty.

Zopsss replied:
> In master, this type of condition was hit on a link to an issue (
> 
> https://github.com/checkstyle/checkstyle/blob/2c232e33c8edccfd7f7658917f7f38208f5a3f2e/src/xdocs/google_style.xml#L199
> 
> ).

yes it is used for this line. Pls refer https://github.com/checkstyle/checkstyle/pull/14925#issuecomment-2157832607

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1818,34 +1891,111 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
+        }
+
+        if (!skip) {
+            if (itrConfigs.hasNext()) {
```
Explain chapter testing. My understanding was, 1 test for many checks. We would still have many configs for many checks.

I am just seeing checks for test links here, and no more config links?

rnveach replied:
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a4.8.5.2

rnveach replied:
I am going to resolve this for now.

Zopsss replied:
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a4.8.5.2

the `skip` isn't used for this, please refer to this comment https://github.com/checkstyle/checkstyle/pull/14925#issuecomment-2157832607

rnveach commented:
GitHub, generate website

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1818,34 +1891,111 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
-                .isFalse();
+    private static void validateChapterWiseTesting(Iterator<Node> itrChecks,
+          Iterator<Node> itrConfigs, Set<String> styleChecks, String styleName, String ruleName) {
+        boolean skip = !itrConfigs.hasNext();
```
Rename skip to `hasChecks`. Set it to false for now.

Zopsss replied:
We cannot change this and other suggestions you gave. Pls refer to this comment detailed explanation: https://github.com/checkstyle/checkstyle/pull/14925

Zopsss replied:
ignore my above comment, pushed new changes according to your suggestions

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1818,34 +1891,111 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
+            final boolean moduleIsCheck = href.startsWith("checks/");
+
+            if (!moduleIsConfig && !moduleIsCheck) {
+                skip = true;
```
Remove this skip assignment here.

If we have checks, we must have atleast 1 config link and 1 test link. If we have no checks, we must have no configs or links.

After this if statement and before first assert set `hasChecks` to true. We are validating a check at this point, we so know we have a check. If we have no links, it will be caught during `itrConfigs.next`. If we have links but it is a test link, it will be caught during expected config link check.

Zopsss replied:
done

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1818,34 +1891,111 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
+
+        }
+
+        if (!skip) {
```
Remove this line `if (!skip) {`.

Zopsss replied:
done

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1818,34 +1891,111 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
+
+        if (!skip) {
+            if (itrConfigs.hasNext()) {
+                final Node config = itrConfigs.next();
```
Assert before accessing `itrConfigs` that `hasChecks` is `true`.

Zopsss replied:
done

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1818,34 +1891,111 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
+            }
+            else {
+                assertWithMessage("google_style.xml rule '" + ruleName + "' should have a test url")
+                        .fail();
```
Change this to assert `hasChecks` is `false`.

Zopsss replied:
done

rnveach reviewed (changes requested):
Some improvements. My main complaint is `if (!skip) {` didn't seem proper. It was possibly to basically not validate any test links if it was ever set improperly.

Zopsss commented:
> My main complaint is `if (!skip) {` didn't seem proper.

I guess you're misunderstanding it's purpose. Let me explain the whole flow of the method, that way it will be clear to you:

There is a list of sections of google style: https://github.com/checkstyle/checkstyle/blob/8db57f20d1729eeb519e3f52ef4a6701f66ab87f/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L233

this list contains all the google style sections which follows per-module approach. Though this list does not contains sections like: [2.2](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a2.2), [3.3.4](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a3.3.4), [5.2.4](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a5.2.4), [6.1](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a6.1), [6.3](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a6.3).

We will be removing sections one-by-one from this list as we migrate them to chapter wise testing approach, at the end this list will be empty and removed. I did not included the above mentioned sections as they does not comes under any of the approach ( nor in chapter-wise neither in per-module ) because at the end we want to remove this list, if we keep those sections under in this list then this list cannot be empty and removed. 

In the [validateStyleModules()](https://github.com/checkstyle/checkstyle/blob/8db57f20d1729eeb519e3f52ef4a6701f66ab87f/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1828) method we're checking if the current documentation is sun style? if yes then simply follow per-module testing approach, if the current documentation is google and the current section is not in the list I mentioned above then it means we need to validate chapter-wise approach. So all the sections I mentioned above ( 2.2, 3.3.4, etc ) which are not included in that list will also be counted in the chapter-wise approach. Here is how we're checking this condition:

https://github.com/checkstyle/checkstyle/blob/8db57f20d1729eeb519e3f52ef4a6701f66ab87f/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1832-L1840

Now let's come to the `skip` part:

https://github.com/checkstyle/checkstyle/blob/8db57f20d1729eeb519e3f52ef4a6701f66ab87f/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1922

this is checking if the current section does not have any check link ( like [3.3.4](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a3.3.4), [5.2.4](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a5.2.4), [6.1](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a6.1), [6.3](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a6.3) ) then we will be skipping validation for that section.

It will set `skip` to true if the current section does not have any check link. otherwise it'll set false.

---

https://github.com/checkstyle/checkstyle/blob/8db57f20d1729eeb519e3f52ef4a6701f66ab87f/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1932-L1934

this is checking if the current section has link but it is not valid ( like [2.2](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a2.2) ) then we will be skipping config link validation.

The above condition will be hit only for section 2.2.

---

https://github.com/checkstyle/checkstyle/blob/8db57f20d1729eeb519e3f52ef4a6701f66ab87f/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1968

this is used to check if we need to skip the validation of test links for current section or not. If `skip` is set to true then we need to skip the validation.

If the current section does not have any check link ( like [3.3.4](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a3.3.4), [5.2.4](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a5.2.4), [6.1](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a6.1), [6.3](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a6.3) ) or if it has but the link is not valid ( like [2.2](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a2.2) ) then it helps us to skip the test link validation for them.

rnveach commented:
> The above condition will be hit only for section 2.2.

For your current code, yes, however, 2 things.

1) I am looking to the future and what looks appropriate in the validation. We should expect random links to appear in the column and to not impact the correct validation. This is an HTML website, so we can literally put things in new places we haven't before.
2) [When you expand this to the (old) section 7.1.1 and the likes](https://checkstyle.org/google_style.html#a7.1.1), this is where I believe the validation will start to be funky. Our link finder doesn't look for specific links in the column. [It finds all links including issues, etc.](https://github.com/checkstyle/checkstyle/blob/a9086cbaaa2311fb30fc40494a5fc132a947219b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1772)

> this is checking if the current section has link but it is not valid ( like [2.2](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/8db57f2_2024193251/google_style.html#a2.2) )

"Not valid" is too strong a word. Link is just not a check link. It can be an issue link, link to another part of our website, to a point in the google style, or a JLS link, both of which are technically valid and feasible to have in this document where appropriate. They do appear in this document in random places.

This will (and eventually) hit in other areas like https://checkstyle.org/google_style.html#a4.2 , https://checkstyle.org/google_style.html#a4.4 , and https://checkstyle.org/google_style.html#a4.6.2 .

This is part of where I think your logic falls apart when it is connected with, ...

> if (!moduleIsConfig && !moduleIsCheck) { 
     skip = true; 
> then we will be skipping config link validation.

This is where I have issues with the code as it is.

Just because a link isn't a check doesn't mean we need to skip all further config link validation.

Imagine we have this in a section ([similar to the (old) section 7.1.1](https://checkstyle.org/google_style.html#a7.1.1)):
`````
1st Check Link
Issue Link
2nd Check Link
`````

You will see and process the 1st check link and it's config. However, once you see the issue link, you will basically signal a "stop" and do no validation on test links. You will still process the 2nd check link and it's config (which is why I say skip is slightly confusing), but your ending logic for the test link will never process, though we should, since we saw 2 check links.

This is what led to my statement `It doesn't make sense to skip (test link check) on unrelated content. If we have checks, we must have atleast 1 config link and 1 test link. If we have no checks, we must have no configs or links.`

> I have added the following code which checks if the section does not have any checks then it should also not contain any config/test url links

Which is all surrounded by the `skip` check.

Zopsss commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1786,8 +1858,8 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
-            assertWithMessage(styleName + "_style.xml rule '" + ruleName + "' module '" + moduleName
-                        + "' shouldn't end with 'Check'")
+            assertWithMessage(styleName + "_style.xml rule '" + ruleName + "' module '"
+                    + moduleName + "' shouldn't end with 'Check'")
```
done

Zopsss commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1801,13 +1873,14 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
                     assertWithMessage(styleName + "_style.xml rule '" + ruleName + "' module '"
-                            + moduleName + "' is missing the config link: " + configName).fail();
+                            + moduleName + "' is missing the config link: " + configName)
+                            .fail();
```
done

Zopsss commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -1801,13 +1873,14 @@ private static void validateStyleModules(Set<Node> checks, Set<Node> configs,
-                    .isEqualTo(configName);
+                        + moduleName + "' has mismatched config/test links")
+                        .that(config.getTextContent().trim())
+                        .isEqualTo(configName);
```
done

Zopsss commented:
@rnveach pushed new changes according to your suggestions. Thanks for all the help!

romani reviewed (approved):
ok to merge

rdiachenko reviewed (approved):
lgtm

rdiachenko commented:
rebased
